File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/play/moderation.ts
/**
 * Campaign-scoped moderation reports: any member (including the DM) may
 * submit or read reports; only the DM may resolve one, and only once.
 */

import { getDb } from '../../db.ts';
import type { ApiResult, JsonValue } from '../../types.ts';
import { authenticate, isActor, isApiResult, findCampaign, requireParticipant } from './shared.ts';

type ReportRow = {
  report_id: string;
  target_id: string;
  reason: string;
  status: string;
  reporter: string;
  sequence: number;
  action: string | null;
  note: string | null;
  resolver: string | null;
};

function reportBody(row: ReportRow): JsonValue {
  const body: Record<string, unknown> = {
    report_id: row.report_id,
    target_id: row.target_id,
    reason: row.reason,
    status: row.status,
    reporter: row.reporter,
    sequence: row.sequence,
  };
  if (row.status === 'resolved') {
    body.action = row.action;
    body.note = row.note;
    body.resolver = row.resolver;
  }
  return body as JsonValue;
}

function nextReportSequence(db: ReturnType<typeof getDb>, campaignId: string): number {
  const row = db
    .prepare('SELECT COALESCE(MAX(sequence), 0) AS max_sequence FROM play_campaign_moderation_reports WHERE campaign_id = ?')
    .get(campaignId) as { max_sequence: number };
  return row.max_sequence + 1;
}

export function createModerationReport(authHeader: string | undefined, campaignId: string, body: JsonValue): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const reportId = body.report_id;
  if (typeof reportId !== 'string' || reportId.length === 0) {
    return { status: 400, body: { error: 'report_id must be a non-empty string' } };
  }

  const targetId = body.target_id;
  if (typeof targetId !== 'string' || targetId.length === 0) {
    return { status: 400, body: { error: 'target_id must be a non-empty string' } };
  }

  const reason = body.reason;
  if (typeof reason !== 'string' || reason.length === 0) {
    return { status: 400, body: { error: 'reason must be a non-empty string' } };
  }

  const existing = db
    .prepare('SELECT report_id FROM play_campaign_moderation_reports WHERE campaign_id = ? AND report_id = ?')
    .get(campaignId, reportId);
  if (existing) {
    return { status: 409, body: { error: 'report_id already exists in this campaign' } };
  }

  const sequence = nextReportSequence(db, campaignId);
  db.prepare(
    'INSERT INTO play_campaign_moderation_reports (campaign_id, sequence, report_id, target_id, reason, status, reporter) VALUES (?, ?, ?, ?, ?, ?, ?)',
  ).run(campaignId, sequence, reportId, targetId, reason, 'open', actor.username);

  return {
    status: 201,
    body: reportBody({
      report_id: reportId,
      target_id: targetId,
      reason,
      status: 'open',
      reporter: actor.username,
      sequence,
      action: null,
      note: null,
      resolver: null,
    }),
  };
}

export function listModerationReports(authHeader: string | undefined, campaignId: string): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const rows = db
    .prepare(
      'SELECT report_id, target_id, reason, status, reporter, sequence, action, note, resolver FROM play_campaign_moderation_reports WHERE campaign_id = ? ORDER BY sequence ASC',
    )
    .all(campaignId) as ReportRow[];

  return { status: 200, body: { reports: rows.map((row) => reportBody(row)) } };
}

export function resolveModerationReport(
  authHeader: string | undefined,
  campaignId: string,
  reportId: string,
  body: JsonValue,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  if (campaign.owner !== actor.username) {
    return { status: 403, body: { error: 'only the campaign dm may resolve reports' } };
  }

  const row = db
    .prepare(
      'SELECT report_id, target_id, reason, status, reporter, sequence, action, note, resolver FROM play_campaign_moderation_reports WHERE campaign_id = ? AND report_id = ?',
    )
    .get(campaignId, reportId) as ReportRow | undefined;
  if (!row) {
    return { status: 404, body: { error: 'report not found' } };
  }

  const action = body.action;
  if (action !== 'allow' && action !== 'remove') {
    return { status: 400, body: { error: "action must be exactly 'allow' or 'remove'" } };
  }

  const note = body.note;
  if (typeof note !== 'string' || note.length === 0) {
    return { status: 400, body: { error: 'note must be a non-empty string' } };
  }

  if (row.status === 'resolved') {
    return { status: 409, body: { error: 'report already resolved' } };
  }

  db.prepare(
    "UPDATE play_campaign_moderation_reports SET status = 'resolved', action = ?, note = ?, resolver = ? WHERE campaign_id = ? AND report_id = ?",
  ).run(action, note, actor.username, campaignId, reportId);

  return {
    status: 200,
    body: reportBody({ ...row, status: 'resolved', action, note, resolver: actor.username }),
  };
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T002352Z_pi_kimi-k2p7-code_python-django/dndsite/views/moderation.py
"""Campaign moderation reports."""

from django.views.decorators.csrf import csrf_exempt

from django.http import JsonResponse

from ..db import _create_moderation_report, _get_moderation_report, _list_moderation_reports, _resolve_moderation_report, db_conn
from ..http import bad_request, conflict, not_found, parse_json, require_method
from .common import require_play_campaign_dm_owner, require_play_campaign_member_or_owner


def _report_response(row):
    """Build a moderation report response dict."""
    base = {
        "report_id": row["report_id"],
        "target_id": row["target_id"],
        "reason": row["reason"],
        "status": row["status"],
        "reporter": row["reporter"],
        "sequence": row["sequence"],
    }
    if row["status"] == "resolved":
        base.update({
            "action": row["action"],
            "note": row["note"],
            "resolver": row["resolver"],
        })
    return base


def moderation_reports(request, id):
    """Dispatch GET/POST for /v1/play/campaigns/{id}/moderation/reports."""
    if request.method == "POST":
        return submit_moderation_report(request, id)
    if request.method == "GET":
        return list_moderation_reports(request, id)
    return require_method(request, "GET", "POST")


@csrf_exempt
def submit_moderation_report(request, id):
    """POST /v1/play/campaigns/{id}/moderation/reports"""
    err = require_method(request, "POST")
    if err is not None:
        return err

    actor, campaign, err = require_play_campaign_member_or_owner(request, id)
    if err is not None:
        return err

    body = parse_json(request)
    if body is None:
        return bad_request("invalid request")

    report_id = body.get("report_id")
    target_id = body.get("target_id")
    reason = body.get("reason")
    if not isinstance(report_id, str) or not report_id:
        return bad_request("report_id is required")
    if not isinstance(target_id, str) or not target_id:
        return bad_request("target_id is required")
    if not isinstance(reason, str) or not reason:
        return bad_request("reason is required")

    try:
        with db_conn() as conn:
            sequence = _create_moderation_report(
                conn, id, report_id, target_id, reason, actor["username"]
            )
            row = _get_moderation_report(conn, id, report_id)
    except Exception:
        return conflict("report already exists")

    return JsonResponse(_report_response(row), status=201)


def list_moderation_reports(request, id):
    """GET /v1/play/campaigns/{id}/moderation/reports"""
    err = require_method(request, "GET")
    if err is not None:
        return err

    actor, campaign, err = require_play_campaign_member_or_owner(request, id)
    if err is not None:
        return err

    with db_conn() as conn:
        rows = _list_moderation_reports(conn, id)

    return JsonResponse({"reports": [_report_response(row) for row in rows]})


@csrf_exempt
def resolve_moderation_report(request, id, report_id):
    """PUT /v1/play/campaigns/{id}/moderation/reports/{report_id}/resolution"""
    err = require_method(request, "PUT")
    if err is not None:
        return err

    actor, campaign, err = require_play_campaign_dm_owner(request, id)
    if err is not None:
        return err

    body = parse_json(request)
    if body is None:
        return bad_request("invalid request")

    action = body.get("action")
    note = body.get("note")
    if action not in ("allow", "remove"):
        return bad_request("action must be allow or remove")
    if not isinstance(note, str) or not note:
        return bad_request("note is required")

    with db_conn() as conn:
        row = _get_moderation_report(conn, id, report_id)
        if row is None:
            return not_found("report not found")
        if row["status"] != "open":
            return conflict("report already resolved")
        _resolve_moderation_report(conn, id, report_id, action, note, actor["username"])
        row = _get_moderation_report(conn, id, report_id)

    return JsonResponse(_report_response(row))
